File: .aoc/submitter.py
import os

import requests


def submit(year, day, task, console, answer):
    if not (day == 1 or day == 2):
        console.log("[red]Wrong task given")
        return False

    if not os.path.exists(".aoc/SESSION_TOKEN"):
        console.log("[red]You need to write your session token to the file '.aoc/SESSION_TOKEN' in order to download your input!")
        return False
    tokenFile = open(".aoc/SESSION_TOKEN")
    AOC_SESSION_COOKIE = tokenFile.read()
    tokenFile.close()

    url = f"https://adventofcode.com/{year}/day/{day}/answer"
    post = {
        'level': task,
        'answer': answer
    }
    result = requests.post(url=url, cookies={"session": AOC_SESSION_COOKIE}, data=post)
    if not result.ok:
        console.log("[red]Error while submitting")
        return False
    if result.text.__contains__("That's not the right answer"):
        console.log("[red]That was not the right answer")
    elif result.text.__contains__("You don't seem to be solving the right level"):
        console.log("[green]You already completed this puzzle")
    elif result.text.__contains__("That's the right answer"):
        console.log("[green]That's the right answer")
        return True
    elif result.text.__contains__("You gave an answer too recently"):
        console.log("[red]You gave an answer too recently. Please wait a moment")
    else:
        console.log("[red]Error while submitting")
        console.log(result.text)
    return False

File: aocweb/submitter.go
package aocweb

import (
	"aoc-cli/cli/flags"
	cli "aoc-cli/output"
	"aoc-cli/utils"
	"bytes"
	"fmt"
	"regexp"

	//"io"
	"net/http"

	"github.com/spf13/viper"
)

func postAnswer(day int, year int, task int, solution string) (string, error) {
	cli.ToPrintf("Submitting solution for day %d year %d", day, year).PrintDebug()
	reqBody := []byte(fmt.Sprintf("level=%d&answer=%s", task, solution))
	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("https://adventofcode.com/%d/day/%d/answer", year, day), bytes.NewReader(reqBody))
	if err != nil {
		return "", err
	}

	cookie := viper.GetString(flags.Flags["cookie"].ViperKey)
	if cookie == "" {
		return "", utils.AOCCLIError("No cookie set. Please set the cookie using the --cookie flag or the config file")
	}

	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Cookie", fmt.Sprintf("session=%s", cookie))

	return executeRequest(req)
}

func Submit(day int, year int, task int, solution string) error {
	result, err := postAnswer(day, year, task, solution)
	if err != nil {
		return utils.AOCCLIError("Error submitting answer!").DebugInfo("submitter", err.Error())
	}

	wholeArticle := regexp.MustCompile("(?ms)<article(.*)</article>").FindString(result)
	wholeArticle = replaceTagRegex(wholeArticle, "</?article>", "")
	wholeArticle = replaceTagRegex(wholeArticle, "</?p>", "")

	if (regexp.MustCompile("That's the right answer")).MatchString(wholeArticle) {
		return nil
	}
	if (regexp.MustCompile("You don't seem to be solving the right level")).MatchString(wholeArticle) {
		return utils.AOCCLIErrorf("Already solved task %d for day %d of %d", task, day, year)
	}
	if (regexp.MustCompile("That's not the right answer")).MatchString(wholeArticle) {
		return utils.AOCCLIErrorf("Wrong answer submitted.\nYour answer: %s", solution)
	}
	if (regexp.MustCompile("You gave an answer too recently")).MatchString(wholeArticle) {
		timeToWait := regexp.MustCompile("You have (.*) left to wait").FindStringSubmatch(wholeArticle)
		formattedTimeToWait := "an unknown timespan"
		if timeToWait == nil {
			timeToWait = regexp.MustCompile("You have (.*) left to wait").FindStringSubmatch(wholeArticle)
			if timeToWait != nil {
				formattedTimeToWait = timeToWait[1]
			}
		} else {
			formattedTimeToWait = timeToWait[1]
		}
		return utils.AOCCLIErrorf("You gave an answer too recently. You have to wait for %s to submit again", formattedTimeToWait)
	}
	return utils.AOCCLIErrorf("Unknown error submitting answer: %s", replaceTagRegex(wholeArticle, "<>", ""))
}
